Hw-44-solution

ipantazi opened the pull request:
Updated version of dependencies in the `pom.xml`.
Added model `User`.
Added the required DTOs: `UserRegistrationRequestDto`, `UserResponseDto`and mapper `UserMapper`. 
Added `AuthenticationController`.
Added FieldMatch annotation.
Added the `RegistrationException`, `DataProcessingException` and `EntityNotFoundException` to the `GlobalExceptionHandler`. Added `UserRepository` and `UserServiceImpl`.
Fix method `save` in the `BookServiceImpl`.
Added method `findByIsbnIncludingDeleted` to the `BookRepository`. 
Added `Pageable`, `produces` and `throw` to the `BookController`. 
Added `users table` to Liquibase

ipantazi committed:
Updated version of dependencies in the `pom.xml`.
Added model `User`.
Added the required DTOs: `UserRegistrationRequestDto`, `UserResponseDto`and mapper `UserMapper`.
Added `AuthenticationController`.
Added FieldMatch annotation.
Added the `RegistrationException`, `DataProcessingException` and `EntityNotFoundException` to the `GlobalExceptionHandler`.
Added `UserRepository` and `UserServiceImpl`.
Fix method `save` in the `BookServiceImpl`.
Added method `findByIsbnIncludingDeleted` to the `BookRepository`.
Added `Pageable`, `produces` and `throw` to the `BookController`.
Added `users table` to Liquibase

Elena-Bruyako commented on src/main/java/mate/academy/onlinebookstore/controller/AuthenticationController.java:
```diff
@@ -0,0 +1,33 @@
+import org.springframework.web.bind.annotation.RestController;
+
+@Tag(name = "User management", description = "Endpoints of management users.")
+@RequestMapping(value = "/api/auth", produces = MediaType.APPLICATION_JSON_VALUE)
```
add `server.servlet.context-path=/api `to application.properties
```suggestion
@RequestMapping("/auth")
```

Elena-Bruyako commented on src/main/java/mate/academy/onlinebookstore/controller/BookController.java:
```diff
@@ -20,24 +23,30 @@
 @Tag(name = "Book management", description = "Endpoints of managing books")
 @RestController
-@RequestMapping("/books")
+@RequestMapping(value = "/books", produces = MediaType.APPLICATION_JSON_VALUE)
```
revert changes

Elena-Bruyako commented on src/main/java/mate/academy/onlinebookstore/controller/BookController.java:
```diff
@@ -20,24 +23,30 @@
+    public List<BookDto> searchBooks(@Valid BookSearchParametersDto searchParameters,
+                                     @ParameterObject Pageable pageable) {
+        List<BookDto> searchedBooks = bookService.search(searchParameters, pageable);
+        if (searchedBooks.isEmpty()) {
```
revert changes

ipantazi committed:
Added hashing password and salt:
added field `salt` to `User`,
added class `HashUtil`,
Added `HashPasswordException` and added method `handleHashPasswordException` to the `CustomGlobalExceptionHandler`,
added method `setHashingPassword` to the `UserMapper`

Elena-Bruyako commented on src/main/java/mate/academy/onlinebookstore/dto/user/UserRegistrationRequestDto.java:
```diff
@@ -0,0 +1,43 @@
+        @Email(message = "Invalid format email.")
+        String email,
+
+        @NotNull(message = "Invalid password. Password shouldn't be null.")
```
Use `@NotBlank `for String

```suggestion
        @NotBlank(message = "Invalid password. Password shouldn't be null.")
```

Elena-Bruyako commented on src/main/java/mate/academy/onlinebookstore/exception/RegistrationException.java:
```diff
@@ -0,0 +1,8 @@
+public class RegistrationException extends RuntimeException {
+    public RegistrationException(String message) {
+        super(message);
+
```
remove empty line
```suggestion
```

Elena-Bruyako commented on src/main/java/mate/academy/onlinebookstore/dto/user/UserRegistrationRequestDto.java:
```diff
@@ -0,0 +1,43 @@
+        @Pattern(
+                regexp = "^[a-zA-Z]{3,50}$", message = "First name must be between 3 "
+                + "and 50 characters and contain only letters."
+        )
```
you can add size or min and max to validate length

Elena-Bruyako commented on src/main/java/mate/academy/onlinebookstore/dto/user/UserRegistrationRequestDto.java:
```diff
@@ -0,0 +1,43 @@
+        @Pattern(regexp = "^[a-zA-Z]{3,50}$", message = "Last name must be between 3 "
+                + "and 50 characters and contain only letters."
+        )
+        String lastName,
```
same here

Elena-Bruyako commented on src/main/java/mate/academy/onlinebookstore/service/user/UserServiceImpl.java:
```diff
@@ -0,0 +1,28 @@
+    @Override
+    public UserResponseDto register(UserRegistrationRequestDto requestDto)
+            throws RegistrationException {
+        if (userRepository.findByEmail(requestDto.email()).isPresent()) {
```
```suggestion
        if (userRepository.existsByEmail(requestDto.email())) {
```

Elena-Bruyako commented on src/main/java/mate/academy/onlinebookstore/service/user/UserServiceImpl.java:
```diff
@@ -0,0 +1,28 @@
+            throw new RegistrationException("Can't register user. Email address already in use");
+        }
+        User user = userMapper.toModel(requestDto);
+        User savedUser = userRepository.save(user);
+        return userMapper.toDto(savedUser);
```
```suggestion
        userRepository.save(user);
        return userMapper.toDto(user);
```

Elena-Bruyako commented on src/main/java/mate/academy/onlinebookstore/validation/FieldMatchValidator.java:
```diff
@@ -0,0 +1,45 @@
+            Object firstValue = firstField.get(value);
+            Object secondValue = secondField.get(value);
+
+            boolean isValid = (firstValue == null && secondValue == null)
```
you can use Objects.equals()

Elena-Bruyako reviewed (changes requested):
see comments

ipantazi committed:
Corrected `@RequestMapping` in the `AuthenticationController` and added `server.servlet.context-path` to `application.properities`.
Reverted changes in `BookController`.
Changed `@NotNul` to `NotBlank` annotations and added `@Size` annotation in `UserRegistrationRequestDto`.
Changed method `findByEmail` to `existsByEmail` in the `UserRepository` and used in `register` method in `UserServiceImpl` class.
Added trow to the method `search` in `BookServiceImpl` class.
Changed check to `Object.equals` in `isValid` method in `FieldMatchValidator` class.

Elena-Bruyako commented on src/main/java/mate/academy/onlinebookstore/mapper/UserMapper.java:
```diff
@@ -0,0 +1,32 @@
+            @MappingTarget
+            User user,
+            UserRegistrationRequestDto userRegistrationRequestDto) {
+        user.setSalt(HashUtil.getSalt());
```
you will encode password in the next task, so you can remove setHashingPassword method

Elena-Bruyako commented on src/main/java/mate/academy/onlinebookstore/util/HashUtil.java:
```diff
@@ -0,0 +1,36 @@
+import java.security.SecureRandom;
+import mate.academy.onlinebookstore.exception.HashPasswordException;
+
+public class HashUtil {
```
remove

Elena-Bruyako commented on src/main/java/mate/academy/onlinebookstore/validation/FieldMatchValidator.java:
```diff
@@ -0,0 +1,45 @@
+                        .addConstraintViolation();
+            }
+            return isValid;
+        } catch (final Exception e) {
```
why final Exception?

Elena-Bruyako commented on src/main/java/mate/academy/onlinebookstore/model/User.java:
```diff
@@ -0,0 +1,37 @@
+    private String password;
+
+    @Column(nullable = false)
+    private byte[] salt;
```
remove

Elena-Bruyako commented on src/main/resources/db/changelog/changes/02-create-users-table.yaml:
```diff
@@ -0,0 +1,44 @@
+                  constraints:
+                    nullable: false
+              - column:
+                  name: salt
```
remove

Elena-Bruyako commented on src/main/resources/db/changelog/changes/02-create-users-table.yaml:
```diff
@@ -0,0 +1,44 @@
+                    nullable: false
+              - column:
+                  name: password
+                  type: varchar(255)
```
check that size the same as in DTO (for other fields also)

Elena-Bruyako reviewed (changes requested):
see comments

ipantazi committed:
Deleted class `HashUtil`, method `setHashingPassword` in the interface `UserMapper`, column `salt` in the class `User`.
Corrected size in `databaseChangeLog` and Dto entity.
Deleted `final` in row #41

ipantazi committed:
Corrected pattern in record `CreateBookRequestDto`.

liliia-ponomarenko commented on pom.xml:
```diff
@@ -142,7 +143,6 @@
         </plugins>
     </build>
-</project>
+</project>
```
add one empty line at the end of the file

liliia-ponomarenko commented on src/main/java/mate/academy/onlinebookstore/config/MapperConfig.java:
```diff
@@ -7,7 +7,7 @@
         injectionStrategy = InjectionStrategy.CONSTRUCTOR,
         nullValueCheckStrategy = NullValueCheckStrategy.ALWAYS,
-        implementationPackage = "<PACKAGE_NAME>.impl"
+        implementationPackage = "mate.academy.onlinebookstore.mapper.impl"
```
remove mate academy from the project structure

liliia-ponomarenko commented on src/main/java/mate/academy/onlinebookstore/exception/CustomGlobalExceptionHandler.java:
```diff
@@ -13,11 +13,48 @@
 public class CustomGlobalExceptionHandler extends ResponseEntityExceptionHandler {
+    @ExceptionHandler(HashPasswordException.class)
+    public ResponseEntity<Object> handleHashPasswordException(HashPasswordException ex) {
+        Map<String, Object> body = new LinkedHashMap<>();
+        body.put("timestamp", LocalDateTime.now());
+        body.put("status", HttpStatus.INTERNAL_SERVER_ERROR.value());
+        body.put("message", ex.getMessage());
```
you can move it to the private method and simplify your methods

liliia-ponomarenko commented on src/main/java/mate/academy/onlinebookstore/service/book/BookServiceImpl.java:
```diff
@@ -23,23 +28,37 @@ public class BookServiceImpl implements BookService {
     public BookDto save(CreateBookRequestDto bookRequestDto) {
+        Optional<Book> existingBook = bookRepository.findByIsbnIncludingDeleted(
+                bookRequestDto.isbn());
+        if (existingBook.isPresent()) {
+            Book book = existingBook.get();
+            if (book.isDeleted()) {
+                throw new DataProcessingException("Can't save book. A book with this ISBN: "
+                        + bookRequestDto.isbn() + " already exists but is marked as deleted.");
+            }
+            throw new DataProcessingException("Can't save book. A book with this ISBN: "
+                    + bookRequestDto.isbn() + " already exists in the database.");
+        }
```
does the user need these details? I think it's enough to know that the ISBN was wrong.

liliia-ponomarenko commented on src/main/java/mate/academy/onlinebookstore/service/book/BookServiceImpl.java:
```diff
@@ -23,23 +28,37 @@ public class BookServiceImpl implements BookService {
+        List<BookDto> searchedBooks = bookRepository.findAll(bookSpecification).stream()
                 .map(bookMapper::toBookDto)
                 .toList();
+        if (searchedBooks.isEmpty()) {
+            throw new ResponseStatusException(HttpStatus.NOT_FOUND,
+                    "No books found based on the given criteria");
+        }
```
is it expected behavior? you should return the empty collection

liliia-ponomarenko commented on src/main/java/mate/academy/onlinebookstore/service/user/UserServiceImpl.java:
```diff
@@ -0,0 +1,27 @@
+    public UserResponseDto register(UserRegistrationRequestDto requestDto)
+            throws RegistrationException {
+        if (userRepository.existsByEmail(requestDto.email())) {
+            throw new RegistrationException("Can't register user. Email address already in use");
```
Let’s  pass the email parameter

liliia-ponomarenko commented on src/main/resources/db/changelog/changes/02-create-users-table.yaml:
```diff
@@ -0,0 +1,39 @@
+databaseChangeLog:
+  - changeSet:
+      id: create-users-table
+      author: mateacademy
```
Don’t  use mate academy in your project

liliia-ponomarenko reviewed (changes requested):
Good job! Let’s fix a few comments ;)

ipantazi committed:
Added one empty line at the end of the `pom.xml`.
Removed `mate academy` from the project structure.
Created private method `bodyBuilder()` in the `CustomGlobalExceptionHandler`.
Corrected methods `search()` and `save()` in the `BookServiceImpl`, changed method `findByIsbnIncludingDeleted()` to `existsByIsbnIncludingDeleted()` in the `BookRepository`.

ipantazi committed:
Added one empty line at the end of the `pom.xml`.
Removed `mate academy` from the project structure.
Created private method `bodyBuilder()` in the `CustomGlobalExceptionHandler`.
Corrected methods `search()` and `save()` in the `BookServiceImpl`, changed method `findByIsbnIncludingDeleted()` to `existsByIsbnIncludingDeleted()` in the `BookRepository`.

liliia-ponomarenko reviewed (approved):
Great! 😊
